Added WinForms Classic tutorial (#14)
DevExpress/XPO: Tutorials/WinForms/Classic/create-persistent-classes-and-connect-xpo-to-database.md

* Add the `Orders` and `Customer` properties to create a [relationship](https://docs.devexpress.com/XPO/2041/concepts/relationships-between-objects) between the **Customer** and **Order** classes.
    ```csharp
    // Customer.cs
    [Association("CustomerOrders")]
    public XPCollection<Order> Orders {
        get { return GetCollection<Order>(nameof(Orders)); }
    }
    
    // Order.cs
    private Customer fCustomer;
    [Association("CustomerOrders")]
    public Customer Customer {
        get { return fCustomer; }
        set { SetPropertyValue(nameof(Customer), ref fCustomer, value); }
    }
    ```
* Add a reference to the **System.Configuration** assembly. This assembly contains classes used to access connection string settings from the application's configuration file [Connection Strings and Configuration Files](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/connection-strings-and-configuration-files).
* Add the ([ConnectionHelper.cs](/Tutorials/WinForms/Classic/CS/DataAccess/ConnectionHelper.cs)) file to your project. 

>   >This code initializes the [Data Access Layer](https://docs.devexpress.com/XPO/2121/Feature-Center/Connecting-to-a-Data-Store/Data-Access-Layer). Refer to the following article for more information: [Connecting to a Data Store](https://docs.devexpress.com/XPO/2020/feature-center/connecting-to-a-data-store).

Reviewer: ```suggestion
  >This code initializes the [Data Access Layer](https://docs.devexpress.com/XPO/2121/Feature-Center/Connecting-to-a-Data-Store/Data-Access-Layer). Refer to the [Connecting to a Data Store](https://docs.devexpress.com/XPO/2020/feature-center/connecting-to-a-data-store) article for more information.
```

    **See also:**   
    [XPO Classes Comparison](https://docs.devexpress.com/XPO/3311/concepts/xpo-classes-comparison)
* Add the `FirstName` and `LastName` properties to the `Customer` class. XPO maps these properties to columns in the Customer table.
    ```csharp
    public string FirstName {
        get { return fFirstName; }
        set { SetPropertyValue(nameof(FirstName), ref fFirstName, value); }
    }

    private string fLastName;
    public string LastName {
        get { return fLastName; }
        set { SetPropertyValue(nameof(LastName), ref fLastName, value); }
    }
    ```
    The `SetPropertyValue` method raises the `PropertyChanged` event that is used by data-bound controls to update their display text.
* Add the read-only `ContactName` property. XPO does not map read-only properties to database columns. Use the `FirstName` and `LastName` properties to calculate the `ContactName` property value.
    ```csharp
    public string ContactName {
        get { return string.Concat(FirstName, " ", LastName); }
    }
    ```
* Add the [PersistentAlias](https://docs.devexpress.com/XPO/DevExpress.Xpo.PersistentAliasAttribute) attribute to the `ContactName` property. 

>   >The [PersistentAlias](https://docs.devexpress.com/XPO/DevExpress.Xpo.PersistentAliasAttribute) attribute specifies an expression that substitutes for the property name in a SQL query. The expression passed to the attribute constructor should follow the [Criteria Language Syntax](https://docs.devexpress.com/CoreLibraries/4928/devexpress-data-library/criteria-language-syntax). 

Reviewer: "**that substitutes for**"? Substitutes what for it?

Reviewer: "should follow"? Do you mean 'should use'?

Author: >"that substitutes for"? Substitutes what for it?

Changed to " an expression that XPO should use instead of the property name"

>"should follow"? Do you mean 'should use'?

Should conform to or should be compatible with. I changed this to "Use Criteria Language to build the expression.

The final variant is: "The [PersistentAlias](https://docs.devexpress.com/XPO/DevExpress.Xpo.PersistentAliasAttribute) attribute specifies an expression that XPO should use instead of the property name in a SQL query. Use [Criteria Language](https://docs.devexpress.com/CoreLibraries/4928/devexpress-data-library/criteria-language-syntax) to build the expression."

bcda9cf74119c0520619f07cec1f14e0822d0825

  [Order.cs](/Tutorials/WinForms/Classic/CS/DataAccess/Order.cs)  
  [Order.vb](/Tutorials/WinForms/Classic/VB/DataAccess/Order.vb)
* Add the `Orders` and `Customer` properties to create a [relationship](https://docs.devexpress.com/XPO/2041/concepts/relationships-between-objects) between the **Customer** and **Order** classes.
    ```csharp
    // Customer.cs
    [Association("CustomerOrders")]
    public XPCollection<Order> Orders {
        get { return GetCollection<Order>(nameof(Orders)); }
    }
    
    // Order.cs
    private Customer fCustomer;
    [Association("CustomerOrders")]
    public Customer Customer {
        get { return fCustomer; }
        set { SetPropertyValue(nameof(Customer), ref fCustomer, value); }
    }
    ```
* Add a reference to the **System.Configuration** assembly. This assembly contains classes used to access connection string settings from the application's configuration file [Connection Strings and Configuration Files](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/connection-strings-and-configuration-files).

> * Add the *ConnectionHelper.cs* file, copy code from the example ([ConnectionHelper.cs](/Tutorials/WinForms/Classic/CS/DataAccess/ConnectionHelper.cs)), and paste it to a new file. 

Reviewer: ```suggestion
* Add the *ConnectionHelper.cs* file, copy code from the [ConnectionHelper.cs](/Tutorials/WinForms/Classic/CS/DataAccess/ConnectionHelper.cs) example, and paste it in a new file. 
```

Author: I changed this sentence to: "* Add the ([ConnectionHelper.cs](/Tutorials/WinForms/Classic/CS/DataAccess/ConnectionHelper.cs)) file to your project.".

ac9357fe065652de642488148e1e67f0d65aaf33

>   >This code can be used without modifications in a real application. Refer to the following article for additional details: [Connecting to a Data Store](https://docs.devexpress.com/XPO/2020/feature-center/connecting-to-a-data-store).

Reviewer: "without modifications in a real application"?

Reviewer: ```suggestion
  >This code can be used in an application without any changes. Refer to the [Connecting to a Data Store](https://docs.devexpress.com/XPO/2020/feature-center/connecting-to-a-data-store) article for more information.
```

Author: >"without modifications in a real application"?

I replaced this sentence with "This code initializes the [Data Access Layer](https://docs.devexpress.com/XPO/2121/Feature-Center/Connecting-to-a-Data-Store/Data-Access-Layer)."

2a11ebbba1b3f716e02f7991650a8a490e88f3dc

* Open the *Program.cs* file and add the `ConnectionHelper.Connect` method call to the `Main` method.

>   >Visual Studio does not create the *Program.vb* file in a VB.NET project, use the `Form1` class constructor, instead.

Reviewer: ```suggestion
  >Visual Studio does not create the *Program.vb* file in a VB.NET project - use the `Form1` class constructor instead.
```

Reviewer: I'm not use why you wrote ", use the `Form1` class constructor, instead", and why you use a comma before instead?

Author: I changed these sentences to ">Visual Studio does not create the *Program.vb* file in a VB.NET project. You can add this code to the `Form1` class constructor before the `InitializeComponent` method".

47b7f58fe3cedf0b3c49ac77a06b2f29a875f0e6

    ```csharp
    [STAThread]
    static void Main() {
        Application.EnableVisualStyles();
        Application.SetCompatibleTextRenderingDefault(false);
        ConnectionHelper.Connect(false);
        Application.Run(new MainForm());
    }
    ```
* Open the application configuration file (*app.config*) and add the connection string. 
    ```xml
    <connectionStrings>
        <add name="XpoTutorial" connectionString="XpoProvider=InMemoryDataStore"/>
    </connectionStrings>
    ```

>     >The [InMemoryDataStore](https://docs.devexpress.com/XPO/DevExpress.Xpo.DB.InMemoryDataStore) provider is for demo and testing purposes. You can use your database server or an embedded database (for example, [SQLite](https://www.sqlite.org/index.html)). XPO supports 14 database engines. Refer to the following articles for details:\

Reviewer: ```suggestion
    >The [InMemoryDataStore](https://docs.devexpress.com/XPO/DevExpress.Xpo.DB.InMemoryDataStore) provider is for demo and testing purposes. You can use your database server or an embedded database (for example, [SQLite](https://www.sqlite.org/index.html)). XPO supports 14 database engines. Refer to the following articles for more information:\
```

# Create persistent classes and connect XPO to the database

&nbsp;&nbsp;&nbsp;
[Back to TOC](../../)
&nbsp;&nbsp;&nbsp;
[Step 2](/connect-data-grid-to-xpo-objects.md)   

* Create a new Windows Forms App project in Visual Studio.
* Add references to these assemblies ([Manage references in a project](https://docs.microsoft.com/en-us/visualstudio/ide/managing-references-in-a-project)):  
  **DevExpress.Data**  
  **DevExpress.Xpo**  
  These assemblies are available under the **Assemblies > Extensions** category in the **Reference Manager**. 
* Create the **DataAccess** folder and add the `Customer` class to it. This class is a model for the Customer table in a database.
* Change the base class to `XPObject` and add a constructor with a `session` parameter:  
    ``` csharp
    using DevExpress.Xpo;

    namespace DxSample.DataAccess {
        public class Customer :XPObject {
            public Customer(Session session) : base(session) { }
        }
    }
    ```
    **See also:**   
    [XPO Classes Comparison](https://docs.devexpress.com/XPO/3311/concepts/xpo-classes-comparison)

> * Add the `FirstName` and `LastName` properties to the `Customer` class. XPO maps these proeprties to columns in the Customer table.

Reviewer: "proeprties"? You should proofread your text.

Author: I proofread all the files: a496223cfeb605e156bdbe699be41915819cbb9c

    ```csharp
    public string FirstName {
        get { return fFirstName; }
        set { SetPropertyValue(nameof(FirstName), ref fFirstName, value); }
    }

    private string fLastName;
    public string LastName {
        get { return fLastName; }
        set { SetPropertyValue(nameof(LastName), ref fLastName, value); }
    }
    ```
    The `SetPropertyValue` method raises the `PropertyChanged` event that is used by data bound controls to update their display text.
* Add the read-only `ContactName` property. This property is not mapped to any column. Use the `FirstName` and `LastName` properties to calcualate the `ContactName` property value.
    ```csharp
    public string ContactName {
        get { return string.Concat(FirstName, " ", LastName); }
    }
    ```

> * Add the [PersistentAliasAttribute](https://docs.devexpress.com/XPO/DevExpress.Xpo.PersistentAliasAttribute) to the `ContactName` property. With this attribute, XPO will be able to include this property in a filter. The expression passed to the attribute constructor should follow the [Criteria Language Syntax](https://docs.devexpress.com/CoreLibraries/4928/devexpress-data-library/criteria-language-syntax). XPO uses the Persistent Alias expression to build a SQL command compatible with the target database.

Reviewer: ```suggestion
* Add the [PersistentAliasAttribute](https://docs.devexpress.com/XPO/DevExpress.Xpo.PersistentAliasAttribute) to the `ContactName` property. This attribute allows XPO to include this property in a filter. The expression passed to the attribute constructor should follow the [Criteria Language Syntax](https://docs.devexpress.com/CoreLibraries/4928/devexpress-data-library/criteria-language-syntax). XPO uses the Persistent Alias expression to build a SQL command that is compatible with the target database.
```

    ```csharp
    [PersistentAlias("concat(FirstName, ' ', LastName)")]
    public string ContactName {
        get { return string.Concat(FirstName, " ", LastName); }
    }
    ```

> * Use the same approach to add the `Order` class with the `ProductName`(String), `OrderDate`(DateTime), and `Freight`(decimal) properties. Check our example code in case of any difficulty:  

Reviewer: "in case of any difficulty"?

Author: 068d1827c0d6d6d5aac4d7f244e578f6d511cb11

  [Order.cs](/Tutorials/WinForms/Classic/CS/DataAccess/Order.cs)  
  [Order.vb](/Tutorials/WinForms/Classic/VB/DataAccess/Order.vb)
* Add the `Orders` and `Customer` properties to create a [relationship](https://docs.devexpress.com/XPO/2041/concepts/relationships-between-objects) between the **Customer** and **Order** classes.
    ```csharp
    // Customer.cs
    [Association("CustomerOrders")]
    public XPCollection<Order> Orders {
        get { return GetCollection<Order>(nameof(Orders)); }
    }
    
    // Order.cs
    private Customer fCustomer;
    [Association("CustomerOrders")]
    public Customer Customer {
        get { return fCustomer; }
        set { SetPropertyValue(nameof(Customer), ref fCustomer, value); }
    }
    ```

> * Add a reference to the **System.Configuration** assembly. This assembly contains classes required to read the connection string settings from the application configuration file [Connection Strings and Configuration Files](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/connection-strings-and-configuration-files).

Reviewer: ```suggestion
* Add a reference to the **System.Configuration** assembly. This assembly contains classes used to access connection string settings from the application's configuration file [Connection Strings and Configuration Files](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/connection-strings-and-configuration-files).
```

* Add the `ConnectionHelper` class and copy the code from the [ConnectionHelper.cs](/Tutorials/WinForms/Classic/CS/DataAccess/ConnectionHelper.cs) ([ConnectionHelper.vb](/Tutorials/WinForms/Classic/VB/DataAccess/ConnectionHelper.vb)) file. In most cases, this generic helper can be used without modifications. The complete documentation about possible connection settings is provided in the XPO documentation: [Connecting to a Data Store](https://docs.devexpress.com/XPO/2020/feature-center/connecting-to-a-data-store).

> * Open the *Program.cs* file and add the `ConnectionHelper.Connect` method call to the `Main` method (the *Program.vb* file does not exist in a VB.NET project by defult, use the `Form1` class constructor, instead):

Reviewer: Firstly, ' by defult' is incorrect, and secondly, do not use 'by default'.

Author: a7c4070d23fb271cea65a538d6dac3c2a81685e0

    ```csharp
    [STAThread]
    static void Main() {
        Application.EnableVisualStyles();
        Application.SetCompatibleTextRenderingDefault(false);
        ConnectionHelper.Connect(false);
        Application.Run(new MainForm());
    }
    ```

> * Open the application configuration file (*app.config*) and add the connection string. The `InMemoryDataStore` provider is used in this tutorial for the sake of simplicity. The complete list of supported XPO providers is provided at [K18445 - How to create a correct connection string for XPO providers](https://www.devexpress.com/Support/Center/Question/Details/K18445).

Reviewer: "for the sake of simplicity"?

Reviewer: " The complete list of supported XPO providers is provided at [K18445 - How to create a correct connection string for XPO providers](https://www.devexpress.com/Support/Center/Question/Details/K18445)." -> Refer to [K18445 - How to create a correct connection string for XPO providers](https://www.devexpress.com/Support/Center/Question/Details/K18445) for a complete list of supported XPO providers.

Author: I refined these texts

e05bb553eec9d3f47ca2e049c9175e8dccefcdb6
191b6f0645c8f7ee66098afcca0f76d7d88237b5
